src/Avalonia.Base/Input/AccessKeyHandler.cs

```diff
@@ -255,5 +280,298 @@ private void MainMenuClosed(object? sender, EventArgs e)
+        }
+
+        private static bool IsTargetable(IInputElement element) =>
+            element is { IsEffectivelyEnabled: true, IsEffectivelyVisible: true };
+
+        private List<IInputElement> CopyMatchingAndPurgeDead(string key)
+        {
+            var matches = new List<IInputElement>(_registrations.Count);
+                
+            // collect live elements with matching key and remove dead elements
+            for (var i = _registrations.Count - 1; i >= 0; i--)
+            {
+                var registration = _registrations[i];
+                var inputElement = registration.GetInputElement();
+                if (inputElement != null)
+                {
+                    if (registration.Key == key)
+                    {
+                        matches.Add(inputElement);
+                    }
+                }
+                else
+                {
+                    _registrations.RemoveAt(i);
+                }
+            }
+
+            // since we collected the elements when iterating from back to front
+            // we need to reverse them to ensure the original order
+            matches.Reverse();
+            
+            return matches;
+        }
+
+        /// <summary>
+        /// Returns targeting information for the given element.
+        /// </summary>
+        /// <param name="element"></param>
+        /// <param name="key"></param>
+        /// <returns>AccessKeyInformation with target for the access key.</returns>
+        private static AccessKeyInformation GetTargetForElement(IInputElement? element, string key)
+        {
+            var info = new AccessKeyInformation();
+            if (element == null)
+                return info;
+
+            var args = new AccessKeyPressedEventArgs(key);
+            element.RaiseEvent(args);
+            info.Target = args.Target;
+
+            return info;
+        }
+
+        /// <summary>
+        /// Checks if the focused element is a descendent of the owner.
+        /// </summary>
+        /// <param name="owner">The owner to check.</param>
+        /// <returns>If focused element is decendant of owner <c>true</c>, otherwise <c>false</c>. </returns>
+        private static bool IsFocusWithinOwner(IInputRoot owner)
+        {
+            var focusedElement = KeyboardDevice.Instance?.FocusedElement;
+            if (focusedElement is not InputElement inputElement)
+                return false;
+
+            var isAncestorOf = owner is Visual root && root.IsVisualAncestorOf(inputElement);
+            return isAncestorOf;
+        }
+
+        /// <summary>
+        /// Sorts the list of targets according to logical ancestors in the hierarchy
+        /// so that child elements, for example within in the content of a tab,
+        /// are processed before the next parent item i.e. the next tab item.
+        /// </summary>
+        private static List<IInputElement> SortByHierarchy(List<IInputElement> targets)
+        {
```

Comment:
Small optimization: return `targets` immediately if we have 0 or 1 items (the very common cases). There's no point in allocating the list and the queue if no sorting effectively occurs.